core/tee/tee_rpmb_fs.c

```diff
@@ -1472,54 +1514,364 @@ static TEE_Result tee_rpmb_get_max_block(uint16_t dev_id, uint32_t *max_block)
+		 * filled. Further keep on extending the buffer up to its max
+		 * size by reading in from RPMB.
+		 */
+		if (fat_entry_dir->num_total_read < RPMB_BUF_MAX_SIZE) {
+			/*
+			 * Read at most as many elements as fit in the buffer
+			 * and no more than the defined number of entries to
+			 * read in at once.
+			 */
+			num_elems_read = MIN(RPMB_BUF_MAX_SIZE -
+				fat_entry_dir->num_total_read,
+				(uint32_t)CFG_RPMB_FS_RD_ENTRIES);
+
+			/*
+			 * Expand the buffer to fit in the additional entries.
+			 */
+			fe = realloc(fe,
+				     (fat_entry_dir->num_buffered +
+				     num_elems_read) * sizeof(*fe));
+			if (!fe) {
+				res = TEE_ERROR_OUT_OF_MEMORY;
+				goto out;
 			}
 
-			/* Move to next fat_entry. */
-			fat_address += sizeof(struct rpmb_fat_entry);
+			fat_entry_dir->rpmb_fat_entry_buf = fe;
+
+			/* Read in to the next free slot in the buffer/cache. */
+			res = tee_rpmb_read(CFG_RPMB_FS_DEV_ID,
+					    fat_address_local,
+					    (uint8_t *)(fe +
+					    fat_entry_dir->num_total_read),
+					    num_elems_read * sizeof(*fe),
+					    NULL, NULL);
+			if (res != TEE_SUCCESS)
+				goto out;
+
+			/* Update number of buffered elements. */
+			fat_entry_dir->num_buffered += num_elems_read;
+		/*
+		 * This happens when we have read as many elements as can
+		 * possibly fit into the buffer.
+		 * As the first part of the buffer serves as our cache,
+		 * we only overwrite the last part that serves as our temporary
+		 * buffer used to iteratively read in entries when the cache
+		 * is full.
+		 */
+		} else {
+			/* This is the temporary buffer maximum size. */
+			num_elems_read = CFG_RPMB_FS_RD_ENTRIES;
+			/* Reset index to beginning of the temporary buffer. */
+			fat_entry_dir->idx_curr = CFG_RPMB_FS_CACHE_SIZE;
+
+			/* Read in elements after the end of the cache. */
+			res = tee_rpmb_read(CFG_RPMB_FS_DEV_ID,
+					    fat_address_local,
+					    (uint8_t *)(fe +
+					    fat_entry_dir->idx_curr),
+					    num_elems_read * sizeof(*fe),
+					    NULL, NULL);
+			if (res != TEE_SUCCESS)
+				goto out;
 		}
 	}
 
+post_read_in:
+	/* Set FAT FS entry address. */
+	if (fat_address)
+		*fat_address = fat_address_local;
+
+	/* Read respective entry from buffer, pointed to by idx_curr. */
+	*fat_entry = fe + fat_entry_dir->idx_curr;
+
+	/* Increment index and counter for entries total read. */
+	fat_entry_dir->idx_curr++;
+	fat_entry_dir->num_total_read++;
+
+	/*
+	 * Indicate last entry was read.
+	 * Ensures we return a zero value for fat_entry on next invocation.
+	 */
+	if (((*fat_entry)->flags & FILE_IS_LAST_ENTRY) != 0)
+		fat_entry_dir->last_reached = true;
+
+	res = TEE_SUCCESS;
+
 out:
-	free(fat_entries);
+	return res;
+}
+
+#if (TRACE_LEVEL >= TRACE_FLOW)
+static void dump_fat(void)
+{
+	struct rpmb_fat_entry *fat_entry = NULL;
+
+	if (fat_entry_dir_init() != TEE_SUCCESS)
+		return;
+
+	while (fat_entry_dir_get_next(&fat_entry, NULL) == TEE_SUCCESS) {
+		if (!fat_entry)
+			break;
```

Comment:
If `fat_entry_dir_get_next()` would return an error when there is nothing to read then this could be written in a more compact way:
```C
while (fat_entry_dir_get_next(&fat_entry, NULL)
	FMSG("flags 0x%x, size %d, address 0x%x, filename '%s'",
	     fe->flags, fe->data_size, fe->start_address, fe->filename);
```

Reply:
Sounds basically good.

This would result in:
- any next entry accessed, no error: return 'TEE_SUCCESS'
- last entry was accessed in previous call, so now return 'TEE_ERROR_ITEM_NOT_FOUND'
- return other errors when the access fails for some reason - other errors than 'TEE_ERROR_ITEM_NOT_FOUND' can occur in fat_entry_dir_get_next.

dump_fat is the simplest function where we did not have had error handling/return values before, as the function just returns void, so the simplification works pretty well.

But, fat_entry_dir_get_next is also called in the more sophisticated functions rpmb_fs_dir_populate and in read_fat, which we have to consider. The point here is that if we iterate until the last element, we now get 'TEE_ERROR_ITEM_NOT_FOUND' instead of a TEE_SUCCESS value combined with a null-pointer for fat_entry.

Thus, I think I still need to specifically read the return value of fat_entry_dir_get_next into a variable (as there might be other errors than 'TEE_ERROR_ITEM_NOT_FOUND'). In case we get another error than 'TEE_ERROR_ITEM_NOT_FOUND', we should return that error as we already do. But, now in cases where there is the 'TEE_ERROR_ITEM_NOT_FOUND' error, we might have to adapt what happens after the while fat_entry_dir_get_next loop, because the case where we iterate through the whole set of FAT FS entries might not be an error per se, but even desired (see read_fat when 'p' is != NULL). But I guess this should be doable.
Just wanted to check back on you again before I make that change.

Reply:
Well indeed the two forms are quite similar once you take into account the more complex case. I'm fine with returning NULL in `fat_entry` to indicate the end.

So, the only change I'm asking is testing `res` like a boolean: `while (!(res = func()))`, `if (res)`  because it's the coding style we have adopted now, and although this particular file has many instances of `if (res != TEE_SUCCESS)`.